druid/src/widget/images.rs

```diff
@@ -0,0 +1,254 @@
+        bc: &BoxConstraints,
+        _data: &T,
+        _env: &Env,
+    ) -> Size {
+        bc.debug_check("Image");
+
+        if bc.is_width_bounded() {
+            bc.max()
+        } else {
+            bc.constrain(self.image_data.get_size())
+        }
+    }
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, _data: &T, _env: &Env) {
+        let offset_matrix =
+            get_affine_from_fill(paint_ctx.size(), self.image_data.get_size(), self.fill);
+
+        // The ImageData's to_piet function does not clip to the image's size
+        // CairoRenderContext is very like druids but with some extra goodies like clip
+        if self.fill == FillStrat::Contain {
+        } else {
+            let clip_rect = Rect::ZERO.with_size(paint_ctx.size());
+            paint_ctx.clip(clip_rect);
+        }
+        self.image_data.to_piet(offset_matrix, paint_ctx);
+    }
+}
+
+/// Stored Image data.
+/// Implements `FromStr` and can be converted to piet draw instructions.
+#[derive(Clone)]
+pub struct ImageData {
+    pixels: Vec<u8>,
+    x_pixels: u32,
+    y_pixels: u32,
+}
+
+impl ImageData {
+    /// Create an empty Image
+    pub fn empty() -> Self {
+        ImageData {
+            pixels: [].to_vec(),
+            x_pixels: 0,
+            y_pixels: 0,
+        }
+    }
+
+    /// This takes the raw binary contents of a file, eg png, jpg and returns a image
+    ///
+    /// If the image crate cant decode a image from the data then a error will be returned
+    pub fn from_data(raw_image: &[u8]) -> Result<Self, Box<dyn Error>> {
+        let dec = image::load_from_memory(raw_image).map_err(|e| e)?.to_rgb();
+
+        let sizeofimage = dec.dimensions();
+        Ok(ImageData {
+            pixels: dec.to_vec(),
+            x_pixels: sizeofimage.0,
+            y_pixels: sizeofimage.1,
+        })
+    }
+
+    /// Get the size in pixels of the contained image.
+    fn get_size(&self) -> Size {
+        Size::new(self.x_pixels as f64, self.y_pixels as f64)
+    }
+
+    /// Convert ImageData into Piet draw instructions
+    pub fn to_piet(&self, offset_matrix: Affine, paint_ctx: &mut PaintCtx) {
+        paint_ctx
+            .with_save(|ctx| {
+                ctx.transform(offset_matrix);
+
+                let im = ctx
+                    .make_image(
+                        self.x_pixels as usize,
+                        self.y_pixels as usize,
+                        &self.pixels,
+                        ImageFormat::Rgb,
+                    )
+                    .unwrap();
+                let rec = Rect::from_origin_size(
+                    (0.0, 0.0),
+                    (self.x_pixels as f64, self.y_pixels as f64),
+                );
+                ctx.draw_image(&im, rec, InterpolationMode::Bilinear);
+
+                Ok(())
+            })
+            .unwrap();
+    }
+}
+
+impl Default for ImageData {
+    fn default() -> Self {
+        ImageData::empty()
+    }
+}
+
+impl FromStr for ImageData {
```

Comment:
oh: I just noticed that this is using `FromStr`, but it's actually expecting the `str` to be a path, and loading it?

`FromStr` is intended explicitly for parsing data from _text_. If we're loading a file and parsing binary data, we should just have a `from_file` constructor on `ImageData` itself, that takes a path (probably the argument type should be `AsRef<Path>` and then does what this fn does.